Add AppNote0017 on when to reuse IDs (#121)
bbc/tams: docs/appnotes/0017-reuse-of-ids.md

Systems should also consider the resource implications of persisting the duplicate flow.
If all of the segments are shared between the Flows, the resource impact may only be at the metadata level and may not justify potential disruption.
If the potential disruption of deleting the duplicate Flow is little/none, it should probably be deleted.

It should be re-iterated that the best approach to content duplication is to avoid it all together.
Duplication makes effective re-use of content harder.
And it is likely an inefficient use of resources.

### Use-cases For Duplication

While we heavily discourage duplication of media under different Source and Flow IDs where at all possible, there a small number of cases where it is appropriate.

#### Clipping

Where a clip is to be created from a section of a Flow, the clip may be the same Flow ID and same Source ID if the clip can be represented by an external system.
That is to say, a system could represent a clip via a Source/Flow ID and a TimeRange.
Alternatively, a new Flow may be created that re-uses the required Objects.
If the clip is time-shifted, it will always be a different Flow with a different Flow ID and different Source ID.

> Weather clips are represented as Flows within TAMS or external to TAMS will be dependent on the need for re-use, and the scalability implication of the increased number of Flow/Source IDs for a given use-case/deployment.

Reviewer: ```suggestion
Whether clips are represented as Flows within TAMS or external to TAMS will be dependent on the need for re-use, and the scalability implication of the increased number of Flow/Source IDs for a given use-case/deployment.
```

# 0017 - When to reuse IDs

## Abstract

The TAMS data model makes use of strong timing and identity, compatible with the [JT-NM Reference Architecture](https://www.jt-nm.org/reference-architecture) and [NMOS IS-04](https://specs.amwa.tv/is-04/releases/v1.3.3/docs/Data_Model_-_Identifier_Mapping.html), to enable re-use of content throughout production workflows.
This Application Note provides guidance on when identities should be re-used, and when new IDs may be needed.
It assumes basic knowledge of the [TAMS Content Model](https://github.com/bbc/tams/blob/main/docs/appnotes/0001-multi-mono-essence-flows-sources.md).

## The Identity Data Model

> The basic principle for identity in TAMS is that if content is identical in media and timing timing it should, in an ideal world, have the same identity.

Reviewer 2: Typo
```suggestion
The basic principle for identity in TAMS is that if content is identical in media and timing it should, in an ideal world, have the same identity.
```

If two media Objects are bit-identical, they should be the same Object and use the same Object ID.
That is to say the same Object should be re-used where possible, and not duplicated.
Implementations may serve the same Object from multiple locations (i.e. via `get_urls` on Flow Segments which use the Object), but the content at all of these locations should be identical.

> [!IMPORTANT]
> It is not currently possible to maintain Object IDs when transferring media between TAMS instances.
> This is because a new Object must be created on the destination store, and this will create a new Object ID.
> It is, however, possible to re-use Object IDs when referencing media held in external storage in another TAMS.

If two Flows have identical data (i.e. Objects) at the same Timestamps, they should be the same Flow and use the same Flow ID and, by extension, the same Source ID.
That is to say the same Flow should be re-used where possible, and not duplicated.

If two Flows are editorially identical, but not technically identical (i.e. are different in technical characteristics alone) they are different Flows but the same Source.
They should have the same Source ID but different Flow IDs.

> If two Flows are technically identical but are timeshifted, the are different Flows with different Flow IDs and different Source IDs.

Reviewer 2: Typo
```suggestion
If two Flows are technically identical but are timeshifted, they are different Flows with different Flow IDs and different Source IDs.
```

If a Flow is ingested from, or outgested to a non-TAMS system and remains editorially identical with no timeshift, but not technically identical, it should be the same Source and should maintain it's Source ID but have different Flow IDs.

If a Flow passes from one system "A" to another system "B" and back to system "A", the Flow and its metadata should be identical at both points in system "A".

It is worth noting that Objects are a TAMS concept.
While they are somewhat analogous to Grains in NMOS, there is no requirement for them to map directly.
The important aspect is the ability of Flows to pass through multiple systems transparently.

## Real-world Implications

It should also be acknowledged that the real world is messy and it might not be possible to meet these recommendations at all times.
Consider where the same media is ingested into two separate systems simultaneously.
Ideally, these ingestors should be configured by the same coordinating system to have the same IDs.
Indeed, dual redundant ingest systems should work in this way.
But there may be cases where such parallel ingests are in entirely separate systems.

### Avoiding Duplication

> Systems may wish to present a list of existing Sources/Flows on setup in such a way that users could identify the content which is being duplicated and either re-use the existing copy (avoiding an un-needed ingest), or ingest with the existing identifiers.

Reviewer 2: Nit: could one or two line breaks / paragraphs be made as this is quite a long paragraph.

It should be re-iterated that the best approach to content duplication is to avoid it all together.
Duplication makes effective re-use of content harder.
And it is likely an inefficient use of resources.

### Use-cases For Duplication

While we heavily discourage duplication of media under different Source and Flow IDs where at all possible, there a small number of cases where it is appropriate.

#### Clipping

Where a clip is to be created from a section of a Flow, the clip may be the same Flow ID and same Source ID if the clip can be represented by an external system.
That is to say, a system could represent a clip via a Source/Flow ID and a TimeRange.
Alternatively, a new Flow may be created that re-uses the required Objects.
If the clip is time-shifted, it will always be a different Flow with a different Flow ID and different Source ID.
Whether clips are represented as Flows within TAMS or external to TAMS will be dependent on the need for re-use, and the scalability implication of the increased number of Flow/Source IDs for a given use-case/deployment.

#### Authorization

> Where authorisation is provided at the Flow level (e.g. by using the mechanism described in ADR-XXXX/AppNote-XXXX) new "clip" Flows may be created to provide access to a subsection of an existing Flow only.

Reviewer 2: Update to this even though the link is not yet possible?
```suggestion
Where authorisation is provided at the Flow level (e.g. by using the mechanism described in ADR-0028/AppNote-0016) new "clip" Flows may be created to provide access to a subsection of an existing Flow only.
```
